lib/Language/Souffle/Interpreter.hs

```diff
@@ -0,0 +1,237 @@
+  run :: Handle prog -> SouffleM ()
+  run (Handle ref) = SouffleM $ do
+    handle <- readIORef ref
+    -- Invoke the souffle binary using parameters, supposing that the facts are placed
+    -- in the factPath, rendering the output into the outputPath
+    callCommand $
+      printf "%s -F%s -D%s -j%d %s"
+        (soufflePath handle)
+        (factPath handle)
+        (outputPath handle)
+        (noOfThreads handle)
+        (datalogExec handle)
+
+  -- | Sets the number of CPU cores this Souffle program should use.
+  setNumThreads :: Handle prog -> Word64 -> SouffleM ()
+  setNumThreads (Handle ref) n = SouffleM $ liftIO $
+    modifyIORef' ref (\h -> h { noOfThreads = n })
+
+  -- | Gets the number of CPU cores this Souffle program should use.
+  getNumThreads :: Handle prog -> SouffleM Word64
+  getNumThreads (Handle ref) = SouffleM $ liftIO $
+    noOfThreads <$> readIORef ref
+
+  -- | Load all facts from files in a certain directory.
+  loadFiles :: Handle prog -> FilePath -> SouffleM ()
+  loadFiles (Handle ref) srcPath = SouffleM $ liftIO $do
+    handle  <- readIORef ref
+    let destPath = factPath handle
+    copyFiles srcPath destPath
+
+  -- | Write out all facts of the program to CSV files
+  --   (as defined in the Souffle program).
+  -- TODO: This should be moved out from this class
+  writeFiles :: Handle prog -> {- FilePath -> -} SouffleM ()
+  writeFiles (Handle ref) {-destPath-} = SouffleM $ liftIO $do
+    handle <- readIORef ref
+    let srcPath = factPath handle
+    copyFiles srcPath "./facts"
+
+  -- | Returns all facts of a program. This function makes use of type inference
+  --   to select the type of fact to return.
+  getFacts :: forall a prog . (Marshal a, Fact a, ContainsFact prog a)
+           => Handle prog -> SouffleM [a]
+  getFacts (Handle ref) = SouffleM $ liftIO $do
+    handle <- readIORef ref
+    let relationName = factName (Proxy :: Proxy a)
+    let factFile = outputPath handle </> relationName <.> "csv"
+    factLines <- readCSVFile factFile
+    let facts = map (runMarshallIT pop) factLines
+    pure facts
+
+  -- | Searches for a fact in a program.
+  --   Returns 'Nothing' if no matching fact was found; otherwise 'Just' the fact.
+  --
+  --   Conceptually equivalent to @List.find (== fact) \<$\> getFacts prog@, but this operation
+  --   can be implemented much faster.
+  findFact :: (Fact a, ContainsFact prog a, Eq a, Marshal a)
+           => Handle prog -> a -> SouffleM (Maybe a)
+  findFact prog fact = find (== fact) <$> getFacts prog
+
+  -- | Adds a fact to the program.
+  addFact :: forall a prog . (Fact a, ContainsFact prog a, Marshal a)
+          => Handle prog -> a -> SouffleM ()
+  addFact (Handle ref) fact = SouffleM $ do
+    handle <- readIORef ref
+    let relationName = factName (Proxy :: Proxy a)
+    let factFile = factPath handle </> relationName <.> "csv"
+    let line = pushMarshallIT (push fact)
+    appendFile factFile $ intercalate "\t" line ++ "\n"
```

Comment:
I still have to figure out how to handle csv separators correctly.

Reply:
I thought a little more about this and I think it's hard to properly implement this (since separator can be specified in Souffle itself). 
For now we can keep it like this, and mention it in the README.